### PR TITLE
Add beforeEmailSent and beforeSmsSent trigger types

### PR DESCRIPTION
**Pull request by blidd-google**
Merges changes from the beforeEmailSent (https://github.com/firebase/firebase-functions/pull/1492) and beforeSmsSent (https://github.com/firebase/firebase-functions/pull/1589) PRs.

### CODE REVIEW

File: src/common/providers/identity.ts [hunk 1]
```diff
@@ -326,26 +334,47 @@ export interface Credential {
...
+export interface BeforeEmailResponse {
+  recaptchaActionOverride?: RecaptchaActionOptions;
+}
+
+/** The handler response type for `beforeSmsSent` blocking events */
+export interface BeforeSmsResponse {
+  recaptchaActionOverride?: RecaptchaActionOptions;
+}
+
+/** The handler response type for beforeCreate blocking events */
```
Comment (egilmorez) [inline src/common/providers/identity.ts:377]: Let's backtick this like the above event type.

File: src/v2/providers/identity.ts [hunk 2]
```diff
@@ -231,59 +223,103 @@ export function beforeUserSignedIn(
...
+ * Handles an event that is triggered before an email is sent to a user.
+ * @param handler - Event handler that is run before an email is sent to a user.
+ */
+export function beforeEmailSent(
+  handler: (event: AuthBlockingEvent) => MaybeAsync<BeforeEmailResponse | void>
+): BlockingFunction;
+
+/**
+ * Handles an event that is triggered before an email is sent to a user.
+ * @param opts - Object containing function options
```
Comment (egilmorez) [inline src/v2/providers/identity.ts:242]: Let's add a period.

File: src/v2/providers/identity.ts [hunk 3]
```diff
@@ -231,59 +223,103 @@ export function beforeUserSignedIn(
...
+ * Handles an event that is triggered before an SMS is sent to a user.
+ * @param handler - Event handler that is run before an SMS is sent to a user.
+ */
+export function beforeSmsSent(
+  handler: (event: AuthBlockingEvent) => MaybeAsync<BeforeSmsResponse | void>
+): BlockingFunction;
+
+/**
+ * Handles an event that is triggered before an SMS is sent to a user.
+ * @param opts - Object containing function options
```
Comment (egilmorez) [inline src/v2/providers/identity.ts:273]: Missing period.

**Review by egilmorez — APPROVED** [review]
LGTM with tiny nits, thanks!

File: src/common/providers/identity.ts [hunk 4]
```diff
@@ -326,26 +334,47 @@ export interface Credential {
   signInMethod: string;
 }
 
+/** Possible types of emails as described by the GCIP backend. */
+export type EmailType = "EMAIL_SIGN_IN" | "PASSWORD_RESET";
+
+/**
+ * The type of SMS message
```
Comment (taeold) [inline src/common/providers/identity.ts:341]: Can we include the comment below to the type comment above so it shows up in the reference docs?

Comment (egilmorez) [reply src/common/providers/identity.ts:341]: With salient details about valid values in the text preferably, thanks!

File: src/common/providers/identity.ts [hunk 5]
```diff
@@ -326,25 +334,50 @@ export interface Credential {
   signInMethod: string;
 }
 
+/**
+ * Possible types of emails as described by the GCIP backend, which can be:
+ * - An sign-in email
```
Comment (egilmorez) [inline src/common/providers/identity.ts:339]: Just "A sign-in email". (not "An")

**Review by egilmorez — APPROVED** [review]
LG!  Miniscule nit in doc string.

**Pull request merged**